crates/wasmtime/src/module.rs

```diff
@@ -108,8 +108,36 @@ struct ModuleInner {
     types: Arc<TypeTables>,
     /// Registered shared signature for the module.
     signatures: Arc<SignatureCollection>,
-    /// a set of memfd images for memories, if any.
-    memfds: Option<Arc<ModuleMemFds>>,
+    /// A set of memfd images for memories, if any. Protected by a
+    /// mutex because module instantiation (hence the need for lazy
+    /// init) may happen for the same module concurrently in multiple
+    /// Stores.
+    memfds: Mutex<LazyMemFds>,
```

Comment:
Given the relative cost of locking a mutex I think it'd be better to use a [`OnceCell`](https://docs.rs/once_cell/latest/once_cell/sync/struct.OnceCell.html) for this which also removes the need for cloning and a custom enum here.

Reply:
Done!